### DISCUSSION TITLE
IP got blocked for any API request

### DISCUSSION THREAD

**Discussion by dayyeeet** [tidal-music/discussions#12] [Report bug]
I just implemented the credentials oauth flow, however, every time i try to make an API request, i receive error code 403/FORBIDDEN and a JSON object in the format:
```json
{
   url: string
}
```

The URL directs to a website informing me that my IP got blocked because of robot like behavior. How can i resolve this issue?

**Comment by aconstas** [comment]
Same thing happens for me. The response from the server is a https://geo.captcha-delivery.com/captcha... link. Haven't had any success with this. Making the requests in the terminal with curl seems to work fine but I can't for the life of me get it to work with fetch.

**Comment by stefanmoro** [comment]
Hey @dayyeeet and @aconstas. Thanks for reaching out and reporting this, we're looking in to the issue.

Could you please supply the entire API request that fails, including headers (with any sensitive info like Authorization header obfuscated), to help us debug.

Thanks

**Reply by dayyeeet** [reply]
This is the code used for the API request:
```ts
async function searchSongs(args: {q?: string, limit?: string, countryCode?: string}, prisma: PrismaClient, token?: MusicProviderAuthToken, ){
    if(!token || !token.token) throw new TRPCError({code: 'BAD_REQUEST', message: 'TidalNoTokenError'})
    if(!args.q) return []
    const res = await fetchTidal('/search', token.token, {
        method: 'GET',
        params: new URLSearchParams({
            query: args.q,
            limit: args.limit as string,
            offset: "0",
            countryCode: args.countryCode ?? 'US',
        })
    })
    //if(!res.ok) throw new TRPCError({code: 'BAD_REQUEST', message: 'TidalSearchError'})
    return []
}

async function fetchTidal(url:string, token: string, args: {method: 'GET' | 'POST', params?: URLSearchParams, headers?: {[index: string]: string}})
{
    return fetch('https://openapi.tidal.com' + url + (args.method === 'GET' && args.params ? `?${args.params.toString()}` : ''), {
        method: args.method,
        headers: {
            'accept': 'application/vnd.tidal.v1+json',
            'Content-Type': 'application/vnd.tidal.v1+json',
            'Authorization': 'Bearer ' + token,
            ...args.headers 
        },
        body: args.method === 'POST' ? args.params : undefined,
    })
}
```

the fetch call in this case reads:
```ts
await fetch(`https://openapi.tidal.com/search?${URLSEARCHPARAMS}`, {
        method: 'GET',
        headers: {
            'accept': 'application/vnd.tidal.v1+json',
            'Content-Type': 'application/vnd.tidal.v1+json',
            'Authorization': 'Bearer ' + TIDAL_TOKEN,
        },
    })
```

**Reply by aliaksei-taliuk** [reply]
@dayyeeet @aconstas It looks like some of the API requests are treated as bot attacks. We're looking for the best solution together with our partners who provide us with a SaaS solution for bot protection. They are asking if we can share your IP addresses so they can properly adjust their models for these kinds of API requests but also whitelist them. You can share your IP address if you need a quick fix (just DM me in GitHub). But at the same time, we'll continue finding better behavior for our generic bot protection mechanism. Will keep you posted!

**Reply by dayyeeet** [reply]
@aliaksei-taliuk unfortunately i can not DM you in GitHub. Can i provide you with a url i received? it contains a hash and a initialCid, maybe your partners can search the IP in the database by this information?

Here is one captcha link: https://geo.captcha-delivery.com/captcha/?initialCid=AHrlqAAAAAMAuF2yRXhR5EwAW0DELA==&cid=7U-HffQr8pI1qeBs-vwFmVGDxEQ-u5_TxOvojB6-rHuwCwlGt6kpJpqudNSeAh-Tb8dp1heacJFX45JdyV~v0QJrnh5E2EIHG78rC4vZg0GiHWwBp99QAlK2nmZRSoKr&referer=http%3A%2F%2Fopenapi.tidal.com%2Fsearch%3Fquery%3Dtest%26limit%3D10%26offset%3D0%26countryCode%3DUS&hash=1F633CDD8EF22541BD6D9B1B8EF13A&t=bv&s=45429&e=a38b68aca321524766ec3a7e37c846540a6382df7317f76e36c3edfafd720bdd

**Reply by aliaksei-taliuk** [reply]
@dayyeeet ok, thanks for the info. I will ask them if it's enough. Meanwhile, feel free to send me a message on my email: aliaksei@tidal.com

**Comment by aconstas** [comment]
@stefanmoro @aliaksei-taliuk Thank you for your prompt response with this issue!
Here is my request that is failing.
```
function callTidalApi() {
    fetch(`https://openapi.tidal.com/search?query=Still%20Woozy&offset=0&limit=10&countryCode=US&popularity=WORLDWIDE`, {
        method: "GET",
        headers: {
            "Authentication": "Bearer: { access token }",
            "Accept": "application/vnd.tidal.v1+json",
            "Content-Type": "application/vnd.tidal.v1+json",
        }
    })
        .then(res => res.json())
        .then(json => console.log(JSON.stringify(json)));
}
```
Here is the response captcha link:
**403 Forbidden**
https://geo.captcha-delivery.com/captcha/?initialCid=AHrlqAAAAAMAUUmLw8WxVnMANFr2bA==&cid=46SOUhXceX_crrLezWTVjfBoF_LOzvpNbFTOLPfjP1Jl77gyjP4YLHIDaU6Z1N_fJ4BislhaFUzZoAaNqH~p_myDMSn0P0VkmQKeUhRtVob08SO6p~q1z5GZHqaStw3E&referer=http%3A%2F%2Fopenapi.tidal.com%2Fartists%2F8669861%2Falbums%3FcountryCode%3DUS%26offset%3D0%26limit%3D10&hash=1F633CDD8EF22541BD6D9B1B8EF13A&t=bv&s=45429&e=a29fb046e40c01b07b2c5c6de4fdc9ad762b7317c5c1fe97e11959048f26cf35

**Reply by aliaksei-taliuk** [reply]
@aconstas I passed all the required data to our partners. Will keep you posted.

**Comment by keybrdist** [comment]
Thanks for looking into this. We are also experiencing this issue.

**Comment by aliaksei-taliuk** [comment]
@aconstas @dayyeeet our partners are still tuning their configurations, but as far as I know, you should be able to send requests now.
@keybrdist most probably your case should be fixed as well. Please try again.
Thanks for your patience.

**Reply by aliaksei-taliuk** [reply]
Our partners let me know that they finally fine-tuned their configuration so now we should have fewer false-positive cases.

**Reply by aconstas** [reply]
Unfortunately, I am still getting a captcha URL alongside the 403 Forbidden in response from the server.

**Comment by keybrdist** [comment]
I'm still getting a blocked message with "captcha" url in the api response.

<img width="694" alt="image" src="https://github.com/tidal-music/discussions/assets/3534206/ea8d8d1e-c0a9-4847-bdb1-385d1d9bd1e6">

My request looks like this.

```php
$url = 'https://openapi.tidal.com/albums/byBarcodeId?barcodeId='.$upc.'&countryCode=US';
$response = Http::withHeaders([
    'Authorization' => 'Bearer '.$access_token,
    'accept' => 'application/vnd.tidal.v1+json',
    'Content-Type' => 'application/vnd.tidal.v1+json',
])->get($url);
```

**Comment by BitesizedLion** [comment]
Encountering the same thing, who the hell puts bot protection on an API?
```
403 Forbidden for URL: https://openapi.tidal.com/albums/87054248?countryCode=US
Response Body:
{"url":"https://geo.captcha-delivery.com/captcha/?initialCid=AHrlqAAAAAMACilCX6K1wQ0AXv-LDQ==&cid=Tdx7vD08G4QArQ-ZivGXi6_vOfkshR192CdyVCrjN-89MAFfr1gwBiW_HtvF4uy2SsFAUj8yT_v6urL3oP_CuELwqO6BfwA1yD6XT1aOOhA_S9rVuiy~UZ1JEcdNQHR&referer=http%3A%2F%2Fopenapi.tidal.com%2Falbums%2F87054248%3FcountryCode%3DUS&hash=1F633CDD8EF22541BD6D9B1B8EF13A&t=fe&s=45429&e=251cacc26c671d9e08a87128cfca5e1b45c05733332310e556dcea1019244d13"}
```

**Comment by aliaksei-taliuk** [comment]
@keybrdist @BitesizedLion please try again, should be fixed now.

**Reply by BitesizedLion** [reply]
Yes, seems to work now.

**Comment by aliaksei-taliuk** [comment]
Closed since we believe the issue has been resolved.
